packages/dds/tree/src/shared-tree-core/transaction.ts

```diff
@@ -245,55 +267,77 @@ export class SquashingTransactionStack<
-			this.setTransactionBranch(transactionBranch);
-			transactionBranch.editor.enterTransaction();
-			return (result) => {
-				assert(this.#transactionBranch !== undefined, 0xa98 /* Expected transaction branch */);
-				this.#transactionBranch.editor.exitTransaction();
-				switch (result) {
-					case TransactionResult.Abort:
-						// When a transaction is aborted, roll back all the transaction's changes on the current branch
-						this.#transactionBranch.removeAfter(startHead);
-						break;
-					case TransactionResult.Commit:
-						// If this was the outermost transaction closing...
-						if (!this.isInProgress()) {
-							if (this.#transactionBranch.getHead() !== startHead) {
-								// ...squash all the new commits on the transaction branch into a new commit on the original branch
-								const removedCommits: GraphCommit<TChange>[] = [];
-								findAncestor(
-									[this.#transactionBranch.getHead(), removedCommits],
-									(c) => c === startHead,
-								);
-								branch.apply(squash(removedCommits));
+		super(
+			// Invoked when an outer transaction starts
+			(): Callbacks => {
+				// Keep track of the commit that each transaction was on when it started
+				// TODO:#8603: This may need to be computed differently if we allow rebasing during a transaction.
+				const startHead = this.activeBranch.getHead();
+				const outerOnPop = onPush?.();
+				const transactionBranch = this.branch.fork(startHead);
+				this.setTransactionBranch(transactionBranch);
+				transactionBranch.editor.enterTransaction();
+				// Invoked when an outer transaction ends
+				const onOuterTransactionPop: OnPop = (result) => {
+					assert(!this.isInProgress(), "The outer transaction should be ending");
+					transactionBranch.editor.exitTransaction();
+					switch (result) {
+						case TransactionResult.Abort:
+							// When a transaction is aborted, roll back all the transaction's changes on the current branch
+							transactionBranch.removeAfter(startHead);
+							break;
+						case TransactionResult.Commit: {
+							// ...squash all the new commits on the transaction branch into a new commit on the original branch
+							const removedCommits: GraphCommit<TChange>[] = [];
+							findAncestor(
+								[transactionBranch.getHead(), removedCommits],
+								(c) => c === startHead,
+							);
+							if (removedCommits.length > 0) {
+								this.branch.apply(squash(removedCommits));
 							}
+							break;
 						}
-						break;
-					default:
-						unreachableCase(result);
-				}
-				if (!this.isInProgress()) {
-					this.#transactionBranch.dispose();
+						default:
+							unreachableCase(result);
+					}
+					transactionBranch.dispose();
 					this.setTransactionBranch(undefined);
-				}
-				onPop?.(result);
-			};
-		});
-
-		this.branch = branch;
+					outerOnPop?.(result);
+				};
+				// Invoked when a nested transaction begins
+				const onNestedTransactionPush: OnPush = () => {
```

Comment:
That's pretty great that you figured out a way to remove this whole special case.